File: Backend/SmartMicrogrid.API/Models/M2/Reservation.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace SmartMicrogrid.API.Models.M2;

public class Reservation
{
    [BsonId, BsonRepresentation(BsonType.ObjectId)] public string? Id { get; set; }
    [BsonElement("prosumerId")] public string ProsumerId { get; set; } = string.Empty;
    [BsonElement("microgridNodeId"), BsonRepresentation(BsonType.ObjectId)] public string MicrogridNodeId { get; set; } = string.Empty;
    [BsonElement("energySlotId"), BsonRepresentation(BsonType.ObjectId)] public string EnergySlotId { get; set; } = string.Empty;
    [BsonElement("energyAmount")] public double EnergyAmount { get; set; }
    [BsonElement("reservationDate")] public DateTime ReservationDate { get; set; }
    [BsonElement("startTime")] public DateTime StartTime { get; set; }
    [BsonElement("endTime")] public DateTime EndTime { get; set; }
    [BsonElement("status")] public string Status { get; set; } = "Pending";
    [BsonElement("createdAt")] public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    [BsonElement("updatedAt")] public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    [BsonElement("statusHistory")] public List<ReservationStatusEvent> StatusHistory { get; set; } = new();
}

public class ReservationStatusEvent
{
    [BsonElement("from")] public string? From { get; set; }
    [BsonElement("to")] public string To { get; set; } = string.Empty;
    [BsonElement("changedBy")] public string ChangedBy { get; set; } = string.Empty;
    [BsonElement("changedAt")] public DateTime ChangedAt { get; set; } = DateTime.UtcNow;
    [BsonElement("reason")] public string? Reason { get; set; }
}

File: Android/SmartMicrogrid.Android/app/src/main/java/com/smartmicrogrid/models/Reservation.kt
package com.smartmicrogrid.models

import com.google.gson.annotations.SerializedName

data class Reservation(
    @SerializedName("id") val id: String = "",
    @SerializedName("prosumerId") val prosumerId: String = "",
    @SerializedName("microgridNodeId") val microgridNodeId: String = "",
    @SerializedName("energySlotId") val energySlotId: String = "",
    @SerializedName("energyAmount") val energyAmount: Double = 0.0,
    @SerializedName("reservationDate") val reservationDate: String = "",
    @SerializedName("startTime") val startTime: String = "",
    @SerializedName("endTime") val endTime: String = "",
    @SerializedName("status") val status: String = "Pending",
    @SerializedName("createdAt") val createdAt: String = "",
    @SerializedName("updatedAt") val updatedAt: String = "",
    @SerializedName("prosumerName") val prosumerName: String? = null,
    @SerializedName("microgridName") val microgridName: String? = null,
    @SerializedName("pricePerUnit") val pricePerUnit: Double = 0.0,
    @SerializedName("totalEstimatedCost") val totalEstimatedCost: Double = 0.0,
    @SerializedName("verificationCode") val verificationCode: String? = null
)

data class CreateReservationRequest(
    @SerializedName("energySlotId") val energySlotId: String,
    @SerializedName("energyAmount") val energyAmount: Double,
    @SerializedName("prosumerId") val prosumerId: String? = null,
    @SerializedName("reservationDate") val reservationDate: String? = null
)

data class ReservationSummary(
    @SerializedName("pendingCount") val pendingCount: Int = 0,
    @SerializedName("approvedFutureCount") val approvedFutureCount: Int = 0,
    @SerializedName("completedThisMonthCount") val completedThisMonthCount: Int = 0,
    @SerializedName("totalEnergyReserved") val totalEnergyReserved: Double = 0.0
)
